### PR TITLE
Code Revamp: Enhanced Performance, Clear Documentation

### PR DESCRIPTION
**Pull request by hanisntsolo**

### CODE REVIEW

File: community-version.py [hunk 1]
```diff
@@ -1,17 +1,64 @@
 #!/usr/bin/python3
-## Community Version
-"""This is class SIMPLEcmd"""
+"""This is the SIMPLEcmd class for ASCII art conversion."""
 
 import os
+import sys
+
+import self
```
Comment (ndeoligence) [inline community-version.py:7]: What's the purpose of importing "self" here? It doesn't seem to being used for its intended purpose in the project

File: requirements.txt [hunk 2]
```diff
@@ -1 +1,3 @@
 Pillow==10.0.1
+
+self~=2020.12.3
```
Comment (ndeoligence) [inline requirements.txt:3]: The self package doesn't seem to be necessary in the above code?

**Review by ndeoligence — COMMENTED** [review]
Hey @hanisntsolo,

Your code might be going over my head here, but it looks like your use of the **self** package is really unnecessary. If so, this might be a big cost, since someone else might confuse it with the **self** used in Python by convention.

I wonder if you didn't install this as a way to resolve some error you got? If that's the case, then it was definitely a mistake - you simply needed to resolve the 'unresolved ref' error. You would've gotten this error (for instance) because of community-version.py:43 (`self.brightness` where `self` would've been unresolved). The editor might give you an option to install and import a package **self**, but that's an uninformed solution - you merely need to resolve the **self** reference. You should find a similar problem in community-version.py:105

**Comment by hanisntsolo** [comment]
> Hey @hanisntsolo,
> 
> 
> 
> Your code might be going over my head here, but it looks like your use of the **self** package is really unnecessary. If so, this might be a big cost, since someone else might confuse it with the **self** used in Python by convention.
> 
> 
> 
> I wonder if you didn't install this as a way to resolve some error you got? If that's the case, then it was definitely a mistake - you simply needed to resolve the 'unresolved ref' error. You would've gotten this error (for instance) because of [community-version.py:43](https://github.com/zero-to-mastery/ascii-art/pull/120/commits/f23352bf5c5e4a15819f208e5a7b6cecb536dec8#diff-7969bfd0923b02d1bc944bea7d3cf38e5464c24162f6a1b01f68ffcc3a56fbcdR43) (`self.brightness` where `self` would've been unresolved). The editor might give you an option to install and import a package **self**, but that's an uninformed solution - you merely need to resolve the **self** reference. You should find a similar problem in [community-version.py:105](https://github.com/zero-to-mastery/ascii-art/pull/120/commits/f23352bf5c5e4a15819f208e5a7b6cecb536dec8#diff-7969bfd0923b02d1bc944bea7d3cf38e5464c24162f6a1b01f68ffcc3a56fbcdR105)

Th guidelines say : 

Make an improvement, it doesnt have to be elaborate

**Comment by kennithnichol** [comment]
@hanisntsolo the criticism in the review is implying that adding this package is in fact **not** an improvement, and may even be detrimental.

I'm looking through these changes myself, and I _think_ you've abstracted the image manipulation functions in to a class?
Have I got that right?

There is a lot of whitespace noise that makes it a bit hard to discern.
Can you summarize what you've done here?

**Comment by hanisntsolo** [comment]
hi @kennithnichol if you want to skip the whitespace noise by default you can approach this approach. 
Ideally while doing a PR review we should always ignore whitespace checks using the below approach/ the view as well whichever suits you best.

image

File: community-version.py (hunk 1 above)
Comment (hanisntsolo) [reply community-version.py:7]: for oop paradigm

**Comment by ndeoligence** [comment]
> > Hey @hanisntsolo,
> > Your code might be going over my head here, but it looks like your use of the **self** package is really unnecessary. If so, this might be a big cost, since someone else might confuse it with the **self** used in Python by convention.
> > I wonder if you didn't install this as a way to resolve some error you got? If that's the case, then it was definitely a mistake - you simply needed to resolve the 'unresolved ref' error. You would've gotten this error (for instance) because of [community-version.py:43](https://github.com/zero-to-mastery/ascii-art/pull/120/commits/f23352bf5c5e4a15819f208e5a7b6cecb536dec8#diff-7969bfd0923b02d1bc944bea7d3cf38e5464c24162f6a1b01f68ffcc3a56fbcdR43) (`self.brightness` where `self` would've been unresolved). The editor might give you an option to install and import a package **self**, but that's an uninformed solution - you merely need to resolve the **self** reference. You should find a similar problem in [community-version.py:105](https://github.com/zero-to-mastery/ascii-art/pull/120/commits/f23352bf5c5e4a15819f208e5a7b6cecb536dec8#diff-7969bfd0923b02d1bc944bea7d3cf38e5464c24162f6a1b01f68ffcc3a56fbcdR105)
> 
> Th guidelines say :
> 
> Make an improvement, it doesnt have to be elaborate

@hanisntsolo, alright. But for what are you using the **self** package in your submission?

File: community-version.py [hunk 3]
```diff
@@ -1,17 +1,64 @@
...
+        """Convert an image to grayscale."""
+        return image.convert('L')
+
+    @staticmethod
+    def map_pixels_to_ascii_chars(image, make_silhouette=False, range_width=25):
+        """Map pixel values to ASCII characters based on a specified range and brightness."""
+        pixels_in_image = list(image.getdata())
+        if make_silhouette:
+            pixels_in_image = [m[3] for m in image.getdata()]
+        adjusted_pixels = [int(pixel * self.brightness) for pixel in pixels_in_image]
```
Comment (ndeoligence) [inline community-version.py:43]: This line introduces a bug. It's related to your use of "self".

File: community-version.py [hunk 4]
```diff
@@ -1,17 +1,64 @@
...
+    def __init__(self, new_width=100, brightness=1.0):
+        self.new_width = new_width
+        self.brightness = brightness
+
+    @staticmethod
+    def scale_image(image):
+        """Resize an image preserving the aspect ratio."""
+        (original_width, original_height) = image.size
+        aspect_ratio = original_height / float(original_width)
+        new_height = int(aspect_ratio * self.new_width)
```
Comment (ndeoligence) [inline community-version.py:28]: This line (and one below) introduces a bug. It's related to your use of "self".

**Review by ndeoligence — CHANGES_REQUESTED** [review]
I had a better look at your code. It introduces bugs at lines 28, 29, and 43 because of your use of `self`. The problem is that you're trying to both make the containing methods static while also referencing instance variables.

Your addition of the **self** package doesn't seem warranted in this submission.

**Comment by kennithnichol** [comment]
> hi @kennithnichol if you want to skip the whitespace noise by default you can approach this approach. Ideally while doing a PR review we should always ignore whitespace checks using the below approach/ the view as well whichever suits you best.
> 
> image

It appears that my diff viewer rendered things very very badly yesterday. I'll blame github. Thanks for your comment, I am aware of the tools provided, and this was more about the line diff noise, rather than 'white space characters'. It looks as I would expect now.

File: community-version.py (hunk 1 above)
Comment (kennithnichol) [reply community-version.py:7]: the 'self' pacakge is a decorator package which lets you do method chaining.

```python
 from self import self

 class CLS:
	@self
	def method(self):
		...

	@self
	def method2(self):
		...
```
```
>>> CLS().method().method2() # jQuery like chain
```

the use of **self** in an OOP sense does not require a package :)

You should be able to remove the package and not affect the functionality.

**Review by kennithnichol — CHANGES_REQUESTED** [review]

**Review by kennithnichol — APPROVED** [review]

**Comment by kennithnichol** [comment]
@hanisntsolo I merged this, because it was complete for the version of the codebase you had.
I have had to revert it (but it will still count as a successful PR) because the master branch base has a lot of changed code since your initial PR. Merging this in erased a lot of things that have since been added.

You're welcome to re-develop your branch with the latest from master branch (you should be able to pull it in from upstream/master if you've set your fork up that way).

It's up to you, but either way -- Thank you 🙌🏻 for working with use to clear your PR.

**Comment by hanisntsolo** [comment]
Now this is what I really liked around working together and that really helped us clear a lot of misunderstanding aka code  and helped us achieve a better perspective at the end of the day.

Ty so much! I'll give it another look once I get some time and rethink what else we can incorporate!
